runtime/src/bank/partitioned_epoch_rewards/distribution.rs

```diff
@@ -42,27 +45,81 @@ struct DistributionResults {
+            return;
+        }
+
+        let (stake_rewards_by_partition, update_to_partitioned) = match &self.epoch_reward_status {
+            EpochRewardStatus::Inactive => {
+                // epoch rewards is inactive, no rewards to distribute
+                unreachable!("shouldn't reach here");
+            }
+            EpochRewardStatus::Calculated(status) => {
+                // epoch rewards have been calculated, but not yet partitioned.
+                // so partition them now.
+                // This should happen only once immediately on the first rewards distribution block, after reward calculation block.
+                let epoch_rewards_sysvar = self.get_epoch_rewards_sysvar();
+                let (stake_rewards_by_partition, partition_us) = measure_us!(status.do_partition(
+                    &epoch_rewards_sysvar.parent_blockhash,
+                    epoch_rewards_sysvar.num_partitions as usize,
+                ));
+                datapoint_info!(
+                    "epoch-rewards-status-update",
+                    ("slot", self.slot(), i64),
+                    ("block_height", height, i64),
+                    ("partition_us", partition_us, i64),
+                    (
+                        "distribution_starting_block_height",
+                        distribution_starting_block_height,
+                        i64
+                    ),
+                );
+
+                (stake_rewards_by_partition, true)
+            }
+            EpochRewardStatus::Partitioned(status) => (
+                // epoch rewards have been partitioned, so use the partitioned rewards
+                status.stake_rewards_by_partition.clone(),
+                false,
+            ),
+        };
+
         let distribution_end_exclusive =
-            distribution_starting_block_height + status.stake_rewards_by_partition.len() as u64;
+            distribution_starting_block_height + stake_rewards_by_partition.len() as u64;
+
         assert!(
             self.epoch_schedule.get_slots_in_epoch(self.epoch)
-                > status.stake_rewards_by_partition.len() as u64
+                > stake_rewards_by_partition.len() as u64
         );
 
         if height >= distribution_starting_block_height && height < distribution_end_exclusive {
             let partition_index = height - distribution_starting_block_height;
+
             self.distribute_epoch_rewards_in_partition(
-                &status.stake_rewards_by_partition,
+                &stake_rewards_by_partition,
                 partition_index,
             );
         }
 
+        if update_to_partitioned {
+            // update epoch reward status to partitioned
+            self.epoch_reward_status =
+                EpochRewardStatus::Partitioned(StartBlockHeightAndPartitionedRewards {
+                    distribution_starting_block_height,
+                    stake_rewards_by_partition,
+                });
+        }
```

Comment:
I would prefer this moved above so that we update right after calculating the partitions.